lib/Conversion/ImportVerilog/FormatStrings.cpp

```diff
@@ -0,0 +1,200 @@
+    // empty literal if no fragments were generated.
+    if (fragments.empty())
+      return Value{};
+    if (fragments.size() == 1)
+      return fragments[0];
+    return builder.create<moore::FormatConcatOp>(loc, fragments).getResult();
+  }
+
+  /// Parse a format string literal and consume and format the arguments
+  /// corresponding to the format specifiers it contains.
+  LogicalResult parseFormat(StringRef format) {
+    bool anyFailure = false;
+    auto onText = [&](auto text) {
+      if (anyFailure)
+        return;
+      emitLiteral(text);
+    };
+    auto onArg = [&](auto specifier, auto offset, auto len,
+                     const auto &options) {
+      if (anyFailure)
+        return;
+      if (failed(emitArgument(specifier, format.substr(offset, len), options)))
+        anyFailure = true;
+    };
+    auto onError = [&](auto, auto, auto, auto) {};
+    slang::SFormat::parse(format, onText, onArg, onError);
+    return failure(anyFailure);
+  }
+
+  /// Emit a string literal that requires no additional formatting.
+  void emitLiteral(StringRef literal) {
+    fragments.push_back(builder.create<moore::FormatLiteralOp>(loc, literal));
+  }
+
+  /// Consume the next argument from the list and emit it according to the given
+  /// format specifier.
+  LogicalResult emitArgument(char specifier, StringRef fullSpecifier,
+                             const FormatOptions &options) {
+    auto specifierLower = std::tolower(specifier);
+
+    // Special handling for format specifiers that consume no argument.
+    if (specifierLower == 'm' || specifierLower == 'l') {
+      return mlir::emitError(loc)
+             << "unsupported format specifier `" << fullSpecifier << "`";
+    }
+
+    // Consume the next argument, which will provide the value to be
+    // formatted.
+    assert(!arguments.empty() && "Slang guarantees correct arg count");
+    const auto &arg = *arguments[0];
+    arguments = arguments.drop_front();
+    auto argLoc = context.convertLocation(arg.sourceRange);
+
+    // Handle the different formatting options.
+    // See IEEE 1800-2017 § 21.2.1.2 "Format specifications".
+    switch (specifierLower) {
+    case 'b':
+      return emitInteger(arg, options, IntFormat::Binary);
+    case 'o':
+      return emitInteger(arg, options, IntFormat::Octal);
+    case 'd':
+      return emitInteger(arg, options, IntFormat::Decimal);
+    case 'h':
+    case 'x':
+      return emitInteger(arg, options,
+                         std::isupper(specifier) ? IntFormat::HexUpper
+                                                 : IntFormat::HexLower);
+
+    case 's':
+      // Simplified handling for literals.
+      if (auto *lit = arg.as_if<slang::ast::StringLiteral>()) {
+        if (options.width)
+          return mlir::emitError(loc)
+                 << "string format specifier with width not supported";
+        emitLiteral(lit->getValue());
+        return success();
+      }
```

Comment:
It only handles literal passed to format string arguments and doesn't accept a string variable like this, correct?
```
string s = "foo";
initial begin
  display("%s", s);
end
```
Maybe it's more flexible if StringLiteralOp takes `moore.string` type SSA value and produces format string value but I think we can revisit this later :+1:

Reply:
Yeah totally! I started out with supporting only literals because they have a simpler special handling. Other cases we'd have to handle with something like `moore.fmt.string` which could take the different flavors of string that can be formatted. SystemVerilog is pretty weird: you can have a `string` type which we need to support, but _unpacked byte arrays_ also count as strings, and string literals like `"foo"` themselves don't have `string` type, but rather generate a `bit [N-1:0]` array of appropriate length. There's also a whole bunch of craziness about taking arbtirary integers and interpreting them as right-aligned ASCII characters. :exploding_head: